package/MDAnalysis/analysis/hydrogenbonds/hbond_analysis.py

```diff
@@ -733,8 +734,11 @@ def _single_frame(self):
         # Remove donor-acceptor pairs between pairs of AtomGroups we are not
         # interested in
         if self.between_ags is not None:
-            tmp_donors, tmp_hydrogens, tmp_acceptors = \
-                self._filter_atoms(tmp_donors, tmp_hydrogens, tmp_acceptors)
+            between_mask = self._filter_atoms(tmp_donors, tmp_acceptors)
+            tmp_donors = tmp_donors[between_mask]
+            tmp_hydrogens = tmp_hydrogens[between_mask]
+            tmp_acceptors = tmp_acceptors[between_mask]
+            d_a_distances = d_a_distances[between_mask]
```

Comment:
a little out of scope, but would it make more sense to filter before the call to `capped_distances`?

Reply:
good point - that would definitely be better, but it's probably not a straightforward change and I imagine there are lots of ways to do it. One way would be to iterate over the pairs of atom selections passed to `between` and calculate donor-acceptor distances for each pair. So e.g. if `between=[["protein", "SOL"], ["protein", "protein"]]`, we would calculate:
- protein donor to SOL acceptor distances
- protein acceptor to SOL donor distances  
- protein donor to protein acceptor distances

and concatenate them. But maybe that should be a separate pr after this fix is in?

Reply:
Separate PR is better — I'd prefer fixing things and knowing that it works and _then_ going for optimizing. I'd be more than happy if someone raises an issue for a performance improvement with a link to this discussion.

Reply:
I raised #4130 — please comment there if you have any follow-ups.